### ISSUE TITLE
strict-comparison false positive (TSLint 5.19.0)

### ISSUE THREAD

**Issue by Akuukis** [palantir/tslint#4838]
# Bug Report

*    TSLint version: 5.19.0
*    TypeScript version: 3.5.3
*    Running TSLint via: VSCode

## TypeScript code being linted

```ts
const MyType = string | boolean

class MyClass<T extends MyType> {
    myMethod(param1: T) {
        if(param1 === true) return 'Yes'  // <-- Cannot compare type 'object' to type 'boolean'. (strict-comparisons)
        if(param1 === false) return 'No'  // <-- Cannot compare type 'object' to type 'boolean'. (strict-comparisons)
        return param1
    }
}
```

## Actual behavior

strict-comparison error.

## Expected behavior

No error.

## Other

Similar to #4774

**Comment by vladeck** [comment]
I would add that the following code does not work as expected (if I understand the rule correctly):

```ts
const a: {foo?: number} = {foo: undefined};
const b: {foo?: number} = {foo: undefined};

// Cannot use '!==' comparator for type 'undefined'. (strict-comparisons)
const eq = a.foo !== b.foo;
```

**Comment by jacekjagiello** [comment]
@adidahiya I'll start working on this one, but I can't reproduce @vladeck case, I'm using Typescript 3.4.5

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is deprecated** and no longer accepting pull requests other than security fixes. See #4534. ☠️ 
We recommend you instead use **[typescript-eslint](https://typescript-eslint.io)** to lint your TypeScript code with **ESLint**. ✅

👋 It was a pleasure open sourcing with you!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
